[GLIB] Many network process crashes when running WPT tests

carlosgcampos opened the pull request:
#### 5b9af92f602ee677d977c4b49768c66f0862a903
<pre>
[GLIB] Many network process crashes when running WPT tests
<a href="https://bugs.webkit.org/show_bug.cgi?id=232629">https://bugs.webkit.org/show_bug.cgi?id=232629</a>

Reviewed by Michael Catanzaro.

Stop using GLib async APIs for network cache IOChannel implementation,
and use the sync API from a thread instead. The async API ends up doing
the job in a thread because file streams are not pollable. The problem
is that the internal GThreadPool tries to change the thread scheduler
setting and it fails when the current thread uses the idle scheduler,
which is the case of disk cache background threads. This reverts the
workaround introduced in 244519@main.

* Source/WTF/wtf/glib/RunLoopSourcePriority.h:
* Source/WebKit/NetworkProcess/cache/NetworkCacheIOChannel.h:
(WebKit::NetworkCache::IOChannel::isOpened const):
* Source/WebKit/NetworkProcess/cache/NetworkCacheIOChannelGLib.cpp:
(WebKit::NetworkCache::IOChannel::IOChannel):
(WebKit::NetworkCache::IOChannel::read):
(WebKit::NetworkCache::IOChannel::write):
(WebKit::NetworkCache::fillDataFromReadBuffer): Deleted.
(WebKit::NetworkCache::inputStreamReadReadyCallback): Deleted.
(WebKit::NetworkCache::IOChannel::readSyncInThread): Deleted.
(WebKit::NetworkCache::outputStreamWriteReadyCallback): Deleted.
* Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp:
(WebKit::NetworkCache::Storage::Storage):
(WebKit::NetworkCache::qosForBackgroundIOQueue): Deleted.

Canonical link: <a href="https://commits.webkit.org/258379@main">https://commits.webkit.org/258379@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9267abca75d7a80248b49bed8e22e2985c1ef7a5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/101662 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/10812 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/34731 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/110996 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/171199 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/11768 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/1723 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/94066 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/108758 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/107442 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/8983 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/92242 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/36556 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/90876 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/23664 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/78552 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/92086 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/4410 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/25160 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/88225 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/2010 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/4485 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/1609 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29361 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/10573 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/44648 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/91131 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6239 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/20354 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3031 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebKit/NetworkProcess/cache/NetworkCacheIOChannelGLib.cpp:
```diff
@@ -145,117 +83,54 @@ void IOChannel::read(size_t offset, size_t size, WTF::WorkQueueBase& queue, Func
+        GRefPtr<GFileInfo> info = adoptGRef(g_file_input_stream_query_info(G_FILE_INPUT_STREAM(m_inputStream.get()), G_FILE_ATTRIBUTE_STANDARD_SIZE, nullptr, nullptr));
+        if (!info) {
+            auto fileSize = g_file_info_get_size(info.get());
+            if (fileSize && static_cast<unsigned>(fileSize) <= std::numeric_limits<size_t>::max()) {
```
`static_cast<unsigned>(fileSize) <= std::numeric_limits<size_t>::max()` is redundant, unless you've found an architecture where a size_t is smaller than an unsigned? ;)

carlosgcampos replied:
Right. The problem is that goffset is gint64, so in 32 bits size_t max can be smaller than the given file size. The comparison without casts gives a warning of different signedness. So, I think the right cast would be `static_cast<guint64>(fileSize)`

carlosgcampos replied:
I replied in the wrong conversation...

mcatanzaro reviewed (approved):
This change makes sense to me regardless of the problem with thread priorities. Not much sense in doing async I/O if we're _already_ on a dedicated thread that exists for no purpose other than this I/O operation, right?

mcatanzaro reviewed (commented):
I don't like that sync I/O on the original thread is required when constructing the IOChannel, but I see that's a preexisting problem. Simple operations like g_file_delete(), g_file_create(), g_file_open_readwrite(), and g_file_read() in the IOChannel constructor can all block the original thread, which is unfortunate. Users with normal disks won't care, but it can be a problem if you're running on NFS, or on an SD card that decides to block on internal recordkeeping, for example.

mcatanzaro commented on Source/WebKit/NetworkProcess/cache/NetworkCacheIOChannelGLib.cpp:
```diff
@@ -145,117 +83,54 @@ void IOChannel::read(size_t offset, size_t size, WTF::WorkQueueBase& queue, Func
+        GRefPtr<GFileInfo> info = adoptGRef(g_file_input_stream_query_info(G_FILE_INPUT_STREAM(m_inputStream.get()), G_FILE_ATTRIBUTE_STANDARD_SIZE, nullptr, nullptr));
+        if (!info) {
+            auto fileSize = g_file_info_get_size(info.get());
+            if (fileSize && fileSize <= static_cast<goffset>(std::numeric_limits<size_t>::max())) {
```
But goffset is signed, so `static_cast<goffset>(std::numeric_limits<size_t>::max())` is going to be negative. Won't the `fileSize <=` check always fail?

mcatanzaro commented on Source/WebKit/NetworkProcess/cache/NetworkCacheIOChannelGLib.cpp:
```diff
@@ -145,117 +81,54 @@ void IOChannel::read(size_t offset, size_t size, WTF::WorkQueueBase& queue, Func
+        GRefPtr<GFileInfo> info = adoptGRef(g_file_input_stream_query_info(G_FILE_INPUT_STREAM(m_inputStream.get()), G_FILE_ATTRIBUTE_STANDARD_SIZE, nullptr, nullptr));
+        if (info) {
+            auto fileSize = g_file_info_get_size(info.get());
+            if (fileSize && static_cast<guint64>(fileSize) <= std::numeric_limits<size_t>::max()) {
```
This makes sense to me.

carlosgcampos committed:
[GLIB] Many network process crashes when running WPT tests
https://bugs.webkit.org/show_bug.cgi?id=232629

Reviewed by Michael Catanzaro.

Stop using GLib async APIs for network cache IOChannel implementation,
and use the sync API from a thread instead. The async API ends up doing
the job in a thread because file streams are not pollable. The problem
is that the internal GThreadPool tries to change the thread scheduler
setting and it fails when the current thread uses the idle scheduler,
which is the case of disk cache background threads. This reverts the
workaround introduced in 244519@main.

* Source/WTF/wtf/glib/RunLoopSourcePriority.h:
* Source/WebKit/NetworkProcess/cache/NetworkCacheIOChannel.h:
(WebKit::NetworkCache::IOChannel::isOpened const):
* Source/WebKit/NetworkProcess/cache/NetworkCacheIOChannelGLib.cpp:
(WebKit::NetworkCache::IOChannel::IOChannel):
(WebKit::NetworkCache::IOChannel::read):
(WebKit::NetworkCache::IOChannel::write):
(WebKit::NetworkCache::fillDataFromReadBuffer): Deleted.
(WebKit::NetworkCache::inputStreamReadReadyCallback): Deleted.
(WebKit::NetworkCache::IOChannel::readSyncInThread): Deleted.
(WebKit::NetworkCache::outputStreamWriteReadyCallback): Deleted.
* Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp:
(WebKit::NetworkCache::Storage::Storage):
(WebKit::NetworkCache::qosForBackgroundIOQueue): Deleted.

Canonical link: https://commits.webkit.org/258379@main
